# Data Wrangling on Auralin Dataset

Auralin dataset is a fabricated clinical trial dataset for the purpose of demonstrating how to undertake a full data wrangling process systematically.

More information about the data...

- This dataset was constructed with the consultation of real doctors to ensure plausibility.
- This clinical trial data for an alternative insulin was inspired and closely mimics this real [clinical trial for a new inhaled insulin called Afrezza](http://care.diabetesjournals.org/content/38/12/2266.long).
- The data quality issues in this dataset mimic real, [common data quality issues in healthcare data](http://media.hypersites.com/clients/1446/filemanager/Articles/DocCenter_Problem_with_data.pdf). These issues impact quality of care, patient registration, and revenue.
- The patients in this dataset were created using this [fake name generator](http://www.fakenamegenerator.com/order.php) and do not include real names, addresses, phone numbers, emails, etc.

---

## Assess

- Visualize by scrolling
- Visualize by programming

In [1]:
pwd

'/Users/alejandrosanz/Downloads'

In [2]:
import os 
os.chdir('projects_on_GitHub/data_wrangling/diabetes_data_wrangling')

In [3]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

pd.set_option('display.max_rows', None)

In [4]:
patients.head(10)

patient_id assigned_sex given_name     surname                     address  \
0           1       female        Zoe     Wellish        576 Brown Bear Drive   
1           2       female     Pamela        Hill   2370 University Hill Road   
2           3         male        Jae      Debord       1493 Poling Farm Road   
3           4         male       Liêm        Phan         2335 Webster Street   
4           5         male        Tim     Neudorf        1428 Turkey Pen Lane   
5           6         male     Rafael       Costa          1140 Willis Avenue   
6           7       female       Mary       Adams            3145 Sheila Lane   
7           8       female     Xiuxiu       Chang  2687 Black Oak Hollow Road   
8           9         male      Dsvid  Gustafsson          1790 Nutter Street   
9          10       female     Sophie     Cabrera           3303 Anmoore Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   
5      Daytona Beach     Florida   32114.0  United States   
6            Burbank          NV   84728.0  United States   
7        Morgan Hill          CA   95037.0  United States   
8        Kansas City          MO   64105.0  United States   
9           New York    New York   10011.0  United States   

                                         contact   birthdate  weight  height  \
0           951-719-9170ZoeWellish@superrito.com   7/10/1976   121.7      66   
1          PamelaSHill@cuvox.de+1 (217) 569-3204    4/3/1967   118.8      66   
2               402-363-6804JaeMDebord@gustr.com   2/19/1980   177.8      71   
3     PhanBaLiem@jourrapide.com+1 (732) 636-8246   7/26/1951   220.9      70   
4                334-515-7487TimNeudorf@cuvox.de   2/18/1928   192.3      27   
5       386-334-5237RafaelCardosoCosta@gustr.com   8/31/1931   183.9      70   
6              775-533-5933MaryBAdams@einrot.com  11/19/1969   146.3      65   
7           XiuxiuChang@einrot.com1 408 778 3236   8/13/1958   158.0      60   
8        816-265-9578DavidGustafsson@armyspy.com    3/6/1937   163.9      66   
9  SophieCabreraIbarra@teleworm.us1 718 795 9124   12/3/1930   194.7      64   

    bmi  
0  19.6  
1  19.2  
2  24.8  
3  31.7  
4  26.1  
5  26.4  
6  24.3  
7  30.9  
8  26.5  
9  33.4

In [5]:
treatments

given_name         surname    auralin    novodra  hba1c_start  \
0             veronika        jindrová  41u - 48u          -         7.63   
1               elliot      richardson          -  40u - 45u         7.56   
2             yukitaka        takenaka          -  39u - 36u         7.68   
3                 skye      gormanston  33u - 36u          -         7.97   
4               alissa          montez          -  33u - 29u         7.78   
5              jasmine           sykes          -  42u - 44u         7.56   
6               sophia          haugen  37u - 42u          -         7.65   
7                eddie          archer  31u - 38u          -         7.89   
8                saber          ménard          -  54u - 54u         8.08   
9                 asia         woźniak  30u - 36u          -         7.76   
10              joseph             day  29u - 36u          -         7.70   
11           kristiina          hyypiä          -  36u - 38u         7.87   
12             roxanne       andreyeva  29u - 38u          -         9.54   
13              gregor            bole          -  47u - 45u         7.61   
14              simone    baumgaertner  27u - 37u          -         7.74   
15                enco          žibrik  55u - 68u          -         7.78   
16             camilla        zaitseva  28u - 37u          -         7.53   
17                gina            cain          -  36u - 36u         7.88   
18          addolorata        lombardi          -  49u - 46u         7.75   
19              khalid        johnsrud          -  54u - 54u         8.35   
20                mile          stanić          -  47u - 48u         7.66   
21               tekla         walczak  29u - 39u          -         7.61   
22         brancaleone           russo  53u - 60u          -         8.61   
23            chiemela       tobeolisa          -  43u - 47u         7.59   
24                isac            berg  31u - 41u          -         9.68   
25              benoît          bonami          -  44u - 43u         9.82   
26              suhaim           rahal          -  49u - 47u         7.94   
27              mizuki           iwata          -  45u - 46u         7.70   
28             clinton          miller  42u - 51u          -         7.79   
29              eugene         mironov  42u - 49u          -         7.81   
30         szilveszter           totth  35u - 39u          -         7.70   
31                 bao             she          -  48u - 45u         9.09   
32               laura     ehrlichmann          -  43u - 40u         7.95   
33              cannan         cabrera          -  40u - 31u         7.94   
34           alexander       mathiesen  47u - 58u          -         7.96   
35               csaba            sági          -  35u - 29u         7.88   
36                 chỉ             lâm  45u - 48u          -         7.68   
37           władysław       wieczorek  24u - 37u          -         7.92   
38            kristján         ingason  44u - 55u          -         7.92   
39           krisztina          magyar          -  32u - 27u         7.52   
40                ásta     grímsdóttir          -  29u - 30u         7.62   
41              mahmud         kadyrov          -  44u - 43u         7.53   
42              marija        grubišić  37u - 43u          -         7.53   
43                jens         poulsen          -  36u - 32u         7.55   
44               sauli      koivuniemi  43u - 47u          -         7.67   
45              zdeněk           synek          -  31u - 31u         7.67   
46             mariana           souza  36u - 42u          -         7.86   
47           jad allah           gaber          -  38u - 41u         7.97   
48                 zak           kelly          -  38u - 38u         7.66   
49             jackson         addison          -  42u - 42u         7.99   
50             albinca         komavec          -  41u - 39u         7.89   
51      

In [6]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [7]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [8]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [9]:

patients.surname.value_counts()

Doe                  6
Jakobsen             3
Taylor               3
Woźniak              2
Liễu                 2
Kadyrov              2
Lâm                  2
Berg                 2
Correia              2
Kowalczyk            2
Dratchev             2
Johnson              2
Hueber               2
Souza                2
Bùi                  2
Lund                 2
Tucker               2
Cabrera              2
Lương                2
Nilsen               2
Silva                2
Batukayev            2
Cindrić              2
Gersten              2
Parker               2
Aranda               2
Ogochukwu            2
Tạ                   2
Collins              2
Grímsdóttir          2
Schiavone            2
Labrosse             1
Dandonneau           1
Teichelmann          1
Koivunen             1
Chinedum             1
Nnonso               1
Vũ                   1
Takenaka             1
Šlezinger            1
Lombardi             1
Wolf                 1
Amari                1
Lindström  

In [10]:
patients.address.value_counts()

123 Main Street                  6
2476 Fulton Street               2
648 Old Dear Lane                2
2778 North Avenue                2
4346 Sussex Court                1
1736 Parrill Court               1
3268 Karen Lane                  1
4148 Callison Lane               1
2917 Golden Ridge Road           1
2578 Tenmile                     1
2386 Linda Street                1
707 Goodwin Avenue               1
57 Norman Street                 1
1160 Taylor Street               1
510 Berry Street                 1
2370 University Hill Road        1
3073 Bedford Street              1
4682 Science Center Drive        1
1507 Woodlawn Drive              1
570 Alpha Avenue                 1
1109 Beechwood Drive             1
3390 Hidden Meadow Drive         1
4852 Rose Avenue                 1
2775 Single Street               1
1826 Poplar Chase Lane           1
1854 Kyle Street                 1
2621 Koontz Lane                 1
4473 Church Street               1
2873 John Calvin Dri

In [11]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
208    121.2
0      121.7
467    122.1
218    122.2
225    122.3
404    123.0
397    123.4
424    123.6
483    123.9
306    124.1
278    124.3
120    124.5
363    124.7
185    125.1
423    125.2
419    126.1
494    126.3
232    126.9
214    126.9
342    127.2
280    128.5
184    128.9
256    128.9
17     129.1
422    129.4
44     129.8
206    129.8
377    130.0
181    130.2
354    130.2
309    130.7
22     130.7
213    131.1
143    131.1
62     131.1
268    132.0
359    132.2
362    132.2
454    132.7
187    133.1
31     133.3
56     133.8
57     134.0
163    134.2
345    134.2
458    134.2
430    135.7

In [12]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [13]:
patients.address.duplicated().sum()

19

In [14]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [15]:
patients[patients[['given_name', 'surname']].duplicated()]

patient_id assigned_sex given_name surname          address      city  \
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

    state  zip_code        country                      contact birthdate  \
229    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
237    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
244    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
251    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
277    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   

     weight  height   bmi  
229   180.0      72  24.4  
237   180.0      72  24.4  
244   180.0      72  24.4  
251   180.0      72  24.4  
277   180.0      72  24.4

In [16]:
patients[patients.address.duplicated()][['given_name', 'surname']]

given_name     surname
29        Jake    Jakobsen
219         Mỹ       Quynh
229       John         Doe
230  Elisabeth     Knudsen
234    Martina   Tománková
237       John         Doe
242       John     O'Brian
244       John         Doe
249   Benjamin      Mehler
251       John         Doe
257        Jin        Kung
264   Wafiyyah      Asfour
269     Flavia  Fiorentino
277       John         Doe
278   Generosa       Cabán
282      Sandy      Taylor
286      Lewis        Webb
296        Chỉ         Lâm
502        Pat     Gersten

In [17]:
treatments.auralin.isnull().sum()

0

In [18]:
treatments.novodra.isnull().sum()

0

In [19]:
# check how many common columns in three tables
all_columns = pd.Series(list(patients) + \
                        list(treatments) + \
                        list(adverse_reactions))

all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [20]:
# # This also works!
# a = set(patients)
# b = set(treatments)
# c = set(adverse_reactions)
# inters = a.intersection(b).intersection(c)
# inters

### Quality Issues
####  `patients` table
- zip code is a float not a string
- zip code has four digits sometimes
- Tim Neudorf height is 27 instead of 72 in
- full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns)
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- multiple phone number formats
- Default John Doe data
- Multiple records for Jakebsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight


#### `treatments` table
- missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- lowercase given names and surnames
- missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (4s mistaken as 9s)
- Nulls represented as dashed (-) in auralin and novodra

#### `adverse_reactions` table
- lowercase given names and surnames



### Tidiness Issues

#### `patients` table
- contact column should be split into phone and email columns separately
 
#### `treatments` table
- three variables in two columns (should be like treatment, start_dose, end_dose)
- `given_name` and `surname` being redundant (Add `id` column to connect with `patients` table)
 
 
#### `adverse_reactions` table
- `given_name` and `surname` being redundant
- This table seems some kind of verbose (should be joined with `treatments` table)


## Clean

**General Process**
- Define
- Code
- Test

In [21]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

#### Completeness Issue (Missing Data) Fixing

**`treatments`: missing records (280 instead of 350)**

___Define___

Import the cut treatments into a DataFrame and concatenate it with the original treatments DataFrame

___Code___

In [22]:
treatments_cut = pd.read_csv('treatments_cut.csv')

In [23]:
treatments_cut.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  42 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


In [24]:
treatments_cut.shape[0] + treatments.shape[0]

350

In [25]:
treatments_clean = pd.concat([treatments_clean, treatments_cut], 
                             ignore_index=True)

___Test___

In [26]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [27]:
treatments_clean.tail()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN

---
**`treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (4s mistaken as 9s)**

___Define___

Recalculate the `hba1c_change` column: `hba1c_start` minus `hba1c_end`. 

___Code___

In [28]:
treatments_clean.hba1c_change = (treatments_clean.hba1c_start - 
                                 treatments_clean.hba1c_end)

___Test___

In [29]:
treatments_clean.hba1c_change.sample(20)

89     0.42
207    0.30
235    0.36
215    0.47
220    0.37
45     0.35
270    0.33
193    0.46
101    0.38
247    0.43
105    0.34
315    0.39
104    0.37
269    0.49
298    0.35
75     0.39
224    0.49
322    0.47
142    0.35
250    0.39
Name: hba1c_change, dtype: float64

#### Tidiness Issue Fixing

**Problem 1**

**`patients` table: `Contact` column contains two variables: phone number and email**

___Define___

Extract the *phone number* and *email* variables from the *contact* column using regular expressions and pandas' `str.extract` method. Drop the *contact* column when done.

___Code___

In [30]:
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

# [a-zA-Z] to signify emails in this dataset all start and end with letters
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop('contact', axis=1)

___Test___

In [31]:
# Confirm contact column is gone
list(patients_clean)
# 'phone_number' in patients_clean.columns.values

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

In [32]:
patients_clean.phone_number.sample(25)

59          252 291 3898
190    +1 (724) 449-6928
464         205-467-9279
304         502-902-6188
151         916-224-7868
186    +1 (401) 485-6384
257                  NaN
27          985-814-7603
228         916-379-7480
321         504 403 4615
56     +1 (706) 497-0891
326         518-379-0603
490         914-636-9304
145    +1 (267) 679-4137
280         210-218-3477
77          859-297-3368
323         513 478 6938
144         260-591-5755
19          530 532 8397
46          724-419-3583
55          702 730 5584
395         336-677-8769
106         765 599 8847
87          320-272-4299
250         410-284-8935
Name: phone_number, dtype: object

In [33]:
# Confirm that no emails start with an integer (regex didn't match for this)
# as ascii code for `0-9` is prior to that of `alphabets`, i.e. ord('1') < ord('a')
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object

**Problem 2**

**`treatments` table: Three variables in two columns (treatment, start dose and end dose) -- Tidiness issue**

___Define___

Melt the *auralin* and *novodra* columns to a *treatment* and a *dose* column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain *start_dose* and *end_dose* columns. Drop the intermediate *dose* column.

___Code___

In [34]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

In [35]:
treatments_clean = treatments_clean.melt(id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'], \
                                         var_name='treatment', \
                                         value_name='dose')

In [36]:
# Test
treatments_clean.sample(25)
# treatments_clean.shape

given_name        surname  hba1c_start  hba1c_end  hba1c_change  \
407  klementyna     sokołowska         7.98       7.53          0.45   
412       hanka          gegič         7.67       7.29          0.38   
634      amanda        ribeiro         7.85       7.47          0.38   
363      gregor           bole         7.61       7.16          0.45   
209        kári     hervinsson         8.09       7.66          0.43   
160      harley         tucker         7.97       7.64          0.33   
119       liisa        seppälä         7.87       7.51          0.36   
200     nicolas       ferreira         7.99       7.72          0.27   
42       marija       grubišić         7.53       7.15          0.38   
247     johanna         dreher         7.77       7.34          0.43   
443      melisa           mata         8.56       8.27          0.29   
570      sinead         reilly         7.76       7.39          0.37   
165      jeramy        hopkins         8.02       7.67          0.35   
189         mee          chung         7.56       7.20          0.36   
284      amanda        ribeiro         7.85       7.47          0.38   
436     ananías       enríquez         7.58       7.07          0.51   
547    angelico       marchesi         7.62       7.18          0.44   
277      mathea        lillebø         9.04       8.67          0.37   
658         una  traustadóttir         8.00       7.50          0.50   
175      marina       glockner         7.88       7.55          0.33   
2      yukitaka       takenaka         7.68       7.25          0.43   
439    stefanie         herman         7.72       7.30          0.42   
346       jakob       jakobsen         7.96       7.51          0.45   
221      torben      mortensen         7.80       7.40          0.40   
419        ivan          fomin         9.12       8.73          0.39   

    treatment       dose  
407   novodra  42u - 41u  
412   novodra  19u - 27u  
634   novodra          -  
363   novodra  47u - 45u  
209   auralin  37u - 43u  
160   auralin  47u - 57u  
119   auralin  36u - 47u  
200   auralin  43u - 51u  
42    auralin  37u - 43u  
247   auralin  51u - 59u  
443   novodra  57u - 57u  
570   novodra          -  
165   auralin  37u - 46u  
189   auralin  42u - 56u  
284   auralin  36u - 44u  
436   novodra  44u - 45u  
547   novodra  58u - 58u  
277   auralin  23u - 36u  
658   novodra  35u - 34u  
175   auralin  35u - 45u  
2     auralin          -  
439   novodra          -  
346   auralin          -  
221   auralin          -  
419   novodra          -

In [37]:
# Only keep the non-null of `dose` records 
treatments_clean = treatments_clean[treatments_clean.dose != "-"]

In [38]:
# Test
treatments_clean.sample(5)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
347      bernd   schneider         7.74       7.44          0.30   auralin   
270       mika  martinsson         7.50       7.17          0.33   auralin   
29      eugene     mironov         7.81       7.48          0.33   auralin   
34   alexander   mathiesen         7.96       7.55          0.41   auralin   
164  felicijan      bubanj         7.85       7.50          0.35   auralin   

          dose  
347  48u - 56u  
270  34u - 43u  
29   42u - 49u  
34   47u - 58u  
164  62u - 75u

In [39]:
# Split `dose` column into two: `dose_start` and `dose_end`
treatments_clean['dose'].str.split(' - ', 1)

0      [41u, 48u]
3      [33u, 36u]
6      [37u, 42u]
7      [31u, 38u]
9      [30u, 36u]
10     [29u, 36u]
12     [29u, 38u]
14     [27u, 37u]
15     [55u, 68u]
16     [28u, 37u]
21     [29u, 39u]
22     [53u, 60u]
24     [31u, 41u]
28     [42u, 51u]
29     [42u, 49u]
30     [35u, 39u]
34     [47u, 58u]
36     [45u, 48u]
37     [24u, 37u]
38     [44u, 55u]
42     [37u, 43u]
44     [43u, 47u]
46     [36u, 42u]
51     [29u, 37u]
52     [57u, 64u]
54     [54u, 67u]
55     [34u, 42u]
58     [29u, 37u]
59     [61u, 67u]
60     [37u, 46u]
61     [30u, 39u]
63     [39u, 46u]
64     [24u, 32u]
66     [32u, 42u]
69     [25u, 32u]
73     [36u, 46u]
78     [35u, 40u]
79     [27u, 36u]
81     [43u, 49u]
82     [33u, 41u]
84     [30u, 41u]
85     [42u, 49u]
87     [31u, 45u]
89     [44u, 55u]
90     [34u, 42u]
95     [39u, 45u]
98     [25u, 31u]
104    [53u, 64u]
109    [41u, 47u]
112    [39u, 52u]
113    [39u, 43u]
115    [32u, 41u]
116    [36u, 44u]
118    [46u, 58u]
119    [36u, 47u]
120    [41

In [40]:
treatments_clean[['dose_start', 'dose_end']] = treatments_clean['dose'].str.split(' - ', expand=True)

# Drop the 'dose' column as it has already been split into `dose_start` and `dose_end`
treatments_clean = treatments_clean.drop('dose', axis=1)

___Test___

In [41]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u  
7        31u      38u  
9        30u      36u

**Problem 3**

**`adverse_reactions` table should be part of the `treatments` table**

___Define___

Merge the *adverse_reaction* column to the `treatments` table, joining on *given name* and *surname*.

___Code___

In [42]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean,
                            on=['given_name', 'surname'], how='left')

___Test___

In [43]:
treatments_clean.sample(10)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
281    cecilie      nilsen         9.63       9.21          0.42   novodra   
52      yasmin       silva         8.88       8.46          0.42   auralin   
245     kamila    pecinová         7.77       7.39          0.38   novodra   
161  štefanija     novosel         7.64       7.17          0.47   auralin   
290     pamela        hill         9.53       9.10          0.43   novodra   
187     benoît      bonami         9.82       9.40          0.42   novodra   
331      meaza      brhane         7.70       7.36          0.34   novodra   
211       hugo     collins         8.00       7.53          0.47   novodra   
1         skye  gormanston         7.97       7.62          0.35   auralin   
312       leon      scholz         7.72       7.29          0.43   novodra   

    dose_start dose_end           adverse_reaction  
281        49u      50u               hypoglycemia  
52         36u      44u                        NaN  
245        54u      51u                        NaN  
161        37u      42u                        NaN  
290        27u      29u                        NaN  
187        44u      43u                        NaN  
331        37u      41u                        NaN  
211        53u      55u                        NaN  
1          33u      36u                        NaN  
312        38u      32u  injection site discomfort

**Problem 4**

Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames -- **Consistency issue**

___Define___

1. Adverse reactions table is no longer needed so ignore that part. 
2. Isolate the patient ID and names in the `patients` table, then convert these names to lower case to join with `treatments`. 
3. Then drop the given name and surname columns in the treatments table (so these being lowercase isn't an issue anymore).

___Code___

In [44]:
# Upper case to lower case
patients_clean['given_name'] = patients_clean['given_name'].apply(lambda x: x.lower())
patients_clean['surname'] = patients_clean['surname'].apply(lambda x: x.lower())


In [45]:
# Test
patients_clean[['given_name', 'surname']].head()

given_name  surname
0        zoe  wellish
1     pamela     hill
2        jae   debord
3       liêm     phan
4        tim  neudorf

In [46]:
# merge `patient_id` into `treatments` table by `given_name` and 'surname' 
treatments_clean = pd.merge(treatments_clean, patients_clean[['patient_id', 'given_name', 'surname']], \
                            on=['given_name', 'surname'])


In [47]:
# Drop `given_name` and `surname` columns
treatments_clean = treatments_clean.drop(['given_name', 'surname'] ,axis=1)

In [48]:
# Test
treatments_clean.sample(20)

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
186         7.59       7.17          0.42   novodra        43u      47u   
148         7.95       7.60          0.35   auralin        46u      57u   
12          8.61       8.18          0.43   auralin        53u      60u   
50          7.85       7.43          0.42   auralin        39u      52u   
111         7.64       7.31          0.33   auralin        52u      61u   
97          9.45       8.94          0.51   auralin        38u      45u   
105         7.68       7.29          0.39   auralin        31u      38u   
334         7.54       7.26          0.28   novodra        37u      40u   
22          7.67       7.37          0.30   auralin        43u      47u   
297         7.92       7.49          0.43   novodra        47u      39u   
151         7.86       7.40          0.46   auralin        26u      39u   
75          7.87       7.57          0.30   auralin        40u      42u   
230         7.73       7.32          0.41   novodra        41u      40u   
220         7.76       7.35          0.41   novodra        34u      32u   
44          7.72       7.30          0.42   auralin        44u      55u   
52          7.64       7.23          0.41   auralin        32u      41u   
277         7.66       7.31          0.35   novodra        39u      44u   
118         7.66       7.29          0.37   auralin        40u      49u   
234         7.93       7.56          0.37   novodra        46u      43u   
152         7.60       7.09          0.51   auralin        22u      31u   

    adverse_reaction  patient_id  
186              NaN         358  
148            cough         270  
12               NaN         316  
50               NaN         371  
111     hypoglycemia          14  
97               NaN         341  
105              NaN          85  
334              NaN          81  
22               NaN         146  
297              NaN         387  
151              NaN         441  
75               NaN         263  
230              NaN         123  
220              NaN         182  
44               NaN         232  
52      hypoglycemia           8  
277              NaN         103  
118              NaN         165  
234              NaN         161  
152              NaN         339

In [49]:
treatments_clean.shape

(349, 8)

---

#### Quality Issue Fixing

**Problem 1**

Zip code is a float not a string and Zip code has four digits sometimes

___Define___

Convert the zip code column's data type from a float to a string using `astype`, remove the '.0' using string slicing, and pad four digit zip codes with a leading 0.

___Code___

In [50]:
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar='0')

# Reconvert NaNs entries that were converted to '0000n' by code above
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)

___Test___

In [51]:
patients_clean.zip_code.head()

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

**Problem 2**

Tim Neudorf height is 27 in instead of 72 in

___Define___

Replace height for rows in the `patients` table that have a height of 27 in (there is only one) with 72 in.

___Code___

In [52]:
patients_clean.height = patients_clean.height.replace(27, 72)

___Test___

In [53]:
# Should be empty
patients_clean[patients_clean.height == 27]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [54]:
# Confirm the replacement worked
patients_clean[patients_clean.surname == 'neudorf']

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        tim  neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      72  26.1   

   phone_number                email  
4  334-515-7487  TimNeudorf@cuvox.de

**Problem 3**

Full state names sometimes, abbreviations other times

___Define___

Apply a function that converts full state name to state abbreviation for California, New York, Illinois, Florida, and Nebraska.

___Code___

In [55]:
# Mapping from full state name to abbreviation
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}

# Function to apply
def abbreviate_state(df):
#     if df['state'] in state_abbrev.keys():
#         abbrev = state_abbrev[df['state']]
#         return abbrev
#     else:
#         return df['state']
    
    return state_abbrev[df['state']] if df['state'] in state_abbrev.keys() else df['state']
    
patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)

___Test___

In [56]:
patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
MA    22
FL    22
PA    18
GA    15
OH    14
MI    13
LA    13
OK    13
NJ    12
VA    11
MS    10
WI    10
MN     9
AL     9
TN     9
IN     9
WA     8
NC     8
KY     8
MO     7
KS     6
NE     6
NV     6
ID     6
CT     5
SC     5
IA     5
ND     4
CO     4
AZ     4
AR     4
RI     4
ME     4
MD     3
SD     3
OR     3
WV     3
DE     3
MT     2
VT     2
DC     2
AK     1
NM     1
NH     1
WY     1
Name: state, dtype: int64

**Problem 4**

Dsvid Gustafsson

___Define___

Replace given name for rows in the `patients` table that have a given name of 'Dsvid' with 'David'.

___Code___

In [60]:
patients_clean[patients_clean['given_name']=='dsvid']

patient_id assigned_sex given_name     surname             address  \
8           9         male      dsvid  gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

In [63]:
patients_clean['given_name'] = patients_clean['given_name'].replace('dsvid', 'david')


___Test___

In [66]:
patients_clean[patients_clean['surname']=='gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      david  gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

**Problem 5**

Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

___Define___

1. Convert assigned sex and state to categorical data types. Zip code data type was already addressed above. 
2. Convert birthdate to datetime data type. 
3. Strip the letter 'u' in start dose and end dose and convert those columns to data type integer.

___Code___


In [67]:
# To category
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

# To datetime
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

# Strip u and to integer
treatments_clean.dose_start = treatments_clean.dose_start.str.strip('u').astype(int)
treatments_clean.dose_end = treatments_clean.dose_end.str.strip('u').astype(int)

___Test___

In [68]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    503 non-null    int64         
 1   assigned_sex  503 non-null    category      
 2   given_name    503 non-null    object        
 3   surname       503 non-null    object        
 4   address       491 non-null    object        
 5   city          491 non-null    object        
 6   state         491 non-null    category      
 7   zip_code      491 non-null    object        
 8   country       491 non-null    object        
 9   birthdate     503 non-null    datetime64[ns]
 10  weight        503 non-null    float64       
 11  height        503 non-null    int64         
 12  bmi           503 non-null    float64       
 13  phone_number  491 non-null    object        
 14  email         491 non-null    object        
dtypes: category(2), datetime64[ns](1), float

In [69]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   hba1c_start       349 non-null    float64
 1   hba1c_end         349 non-null    float64
 2   hba1c_change      349 non-null    float64
 3   treatment         349 non-null    object 
 4   dose_start        349 non-null    int64  
 5   dose_end          349 non-null    int64  
 6   adverse_reaction  35 non-null     object 
 7   patient_id        349 non-null    int64  
dtypes: float64(3), int64(3), object(2)
memory usage: 24.5+ KB


**Problem 6**

Multiple phone number formats

___Define___

1. Strip all " ", "-", "(", ")", and "+" and store each number without any formatting. 
2. Pad the phone number with a 1 if the length of the number is 10 digits (we want country code).

___Code___


In [70]:
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')

___Test___

In [71]:
patients_clean.phone_number.sample(10)

62     16178835967
161    14067596160
82     16783502390
66     17602878295
45     17316322908
337    12088302415
115    17072626503
499    19282844492
18     14067752696
43     13077120508
Name: phone_number, dtype: object

**Problem 7**

Default John Doe data

___Define___

Remove the non-recoverable John Doe records from the patients table as no 'doe' 's treatment records.

___Code___


In [81]:
patients_clean = patients_clean[patients_clean.surname != 'doe']

___Test___

In [82]:
# Should be no Doe records
patients_clean.surname.value_counts()

jakobsen             3
taylor               3
lương                2
souza                2
lund                 2
dratchev             2
cindrić              2
cabrera              2
lâm                  2
liễu                 2
tucker               2
correia              2
nilsen               2
berg                 2
batukayev            2
grímsdóttir          2
ogochukwu            2
collins              2
schiavone            2
parker               2
aranda               2
johnson              2
bùi                  2
kadyrov              2
hueber               2
silva                2
kowalczyk            2
woźniak              2
gersten              2
tạ                   2
carvalho             1
rap                  1
dvořák              1
efrem                1
azuma                1
kung                 1
allaire              1
tansey               1
nicholls             1
isaksson             1
kos                  1
prince               1
bonami               1
mikkelsen  

**Problem 8**

Multiple records for Jakobsen, Gersten, Taylor

___Define___

Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the `patients` table. These are the nicknames, which happen to also not be in the `treatments` table (removing the wrong name would create a consistency issue between the `patients` and `treatments` table). These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.

___Code___


In [90]:
patients_clean[patients_clean.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       jake    jakobsen   648 Old Dear Lane   
219         220         male         mỹ       quynh                 NaN   
230         231       female  elisabeth     knudsen                 NaN   
234         235       female    martina   tománková                 NaN   
242         243         male       john     o'brian                 NaN   
249         250         male   benjamin      mehler                 NaN   
257         258         male        jin        kung                 NaN   
264         265       female   wafiyyah      asfour                 NaN   
269         270       female     flavia  fiorentino                 NaN   
278         279       female   generosa       cabán                 NaN   
282         283       female      sandy      taylor  2476 Fulton Street   
286         287         male      lewis        webb                 NaN   
296         297       female        chỉ         lâm                 NaN   
502         503         male        pat     gersten   2778 North Avenue   

            city state zip_code        country  birthdate  weight  height  \
29   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
219          NaN   NaN      NaN            NaN 1978-04-09   237.8      69   
230          NaN   NaN      NaN            NaN 1976-09-23   165.9      63   
234          NaN   NaN      NaN            NaN 1936-04-07   199.5      65   
242          NaN   NaN      NaN            NaN 1957-02-25   205.3      74   
249          NaN   NaN      NaN            NaN 1951-10-30   146.5      69   
257          NaN   NaN      NaN            NaN 1995-05-17   231.7      69   
264          NaN   NaN      NaN            NaN 1989-11-03   158.6      63   
269          NaN   NaN      NaN            NaN 1937-10-09   175.2      61   
278          NaN   NaN      NaN            NaN 1962-12-16   124.3      69   
282     Rainelle    WV    25962  United States 1960-10-23   206.1      64   
286          NaN   NaN      NaN            NaN 1979-04-01   155.3      68   
296          NaN   NaN      NaN            NaN 1990-05-14   181.1      63   
502         Burr    NE    68324  United States 1954-05-03   138.2      71   

      bmi phone_number                      email  
29   24.4  18458587707  JakobCJakobsen@einrot.com  
219  35.1          NaN                        NaN  
230  29.4          NaN                        NaN  
234  33.2          NaN                        NaN  
242  26.4          NaN                        NaN  
249  21.6          NaN                        NaN  
257  34.2          NaN                        NaN  
264  28.1          NaN                        NaN  
269  33.1          NaN                        NaN  
278  18.4          NaN                        NaN  
282  35.4  13044382648   SandraCTaylor@dayrep.com  
286  23.6          NaN                        NaN  
296  32.1          NaN                        NaN  
502  19.3  14028484923   PatrickGersten@rhyta.com

In [91]:
patients_clean[patients_clean.address.isin(['648 Old Dear Lane', '2476 Fulton Street', '2778 North Avenue'])]

patient_id assigned_sex given_name   surname             address  \
24           25         male      jakob  jakobsen   648 Old Dear Lane   
29           30         male       jake  jakobsen   648 Old Dear Lane   
97           98         male    patrick   gersten   2778 North Avenue   
131         132       female     sandra    taylor  2476 Fulton Street   
282         283       female      sandy    taylor  2476 Fulton Street   
502         503         male        pat   gersten   2778 North Avenue   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
29   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
97          Burr    NE    68324  United States 1954-05-03   138.2      71   
131     Rainelle    WV    25962  United States 1960-10-23   206.1      64   
282     Rainelle    WV    25962  United States 1960-10-23   206.1      64   
502         Burr    NE    68324  United States 1954-05-03   138.2      71   

      bmi phone_number                      email  
24   24.4  18458587707  JakobCJakobsen@einrot.com  
29   24.4  18458587707  JakobCJakobsen@einrot.com  
97   19.3  14028484923   PatrickGersten@rhyta.com  
131  35.4  13044382648   SandraCTaylor@dayrep.com  
282  35.4  13044382648   SandraCTaylor@dayrep.com  
502  19.3  14028484923   PatrickGersten@rhyta.com

In [95]:
# Delete the duplicated ones and keep one-only ccurrence record. As for the 'nan' in address, keep as they used to be.
patients_clean = patients_clean[~(patients_clean.address.duplicated() & patients_clean.address.notnull())]

___Test___

In [96]:
patients_clean[patients_clean.address.isin(['648 Old Dear Lane', '2476 Fulton Street', '2778 North Avenue'])]

patient_id assigned_sex given_name   surname             address  \
24           25         male      jakob  jakobsen   648 Old Dear Lane   
97           98         male    patrick   gersten   2778 North Avenue   
131         132       female     sandra    taylor  2476 Fulton Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
97          Burr    NE    68324  United States 1954-05-03   138.2      71   
131     Rainelle    WV    25962  United States 1960-10-23   206.1      64   

      bmi phone_number                      email  
24   24.4  18458587707  JakobCJakobsen@einrot.com  
97   19.3  14028484923   PatrickGersten@rhyta.com  
131  35.4  13044382648   SandraCTaylor@dayrep.com

In [98]:
patients_clean[patients_clean.surname == 'jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      jakob  jakobsen   648 Old Dear Lane   
432         433       female      karen  jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi phone_number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

In [99]:
patients_clean[patients_clean.surname == 'gersten']

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    patrick  gersten  2778 North Avenue  Burr   

   state zip_code        country  birthdate  weight  height   bmi  \
97    NE    68324  United States 1954-05-03   138.2      71  19.3   

   phone_number                     email  
97  14028484923  PatrickGersten@rhyta.com

In [100]:
patients_clean[patients_clean.surname == 'taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     sandra  taylor  2476 Fulton Street  Rainelle   
426         427         male    rogelio  taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

    phone_number                       email  
131  13044382648    SandraCTaylor@dayrep.com  
426  13054346299  RogelioJTaylor@teleworm.us

**Problem 9**

kgs instead of lbs for Zaitseva weight

___Define___

##### Define
Use [advanced indexing](https://stackoverflow.com/a/44913631) to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

___Code___

In [101]:
patients_clean[patients_clean.surname=='zaitseva']

patient_id assigned_sex given_name   surname              address  \
210         211       female    camilla  zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate  weight  height   bmi  \
210  Wooster    OH    44691  United States 1938-11-26    48.8      63  19.1   

    phone_number                          email  
210  13302022145  CamillaZaitseva@superrito.com

In [115]:
patients_clean.weight.min()

48.8

In [126]:
patients_clean[patients_clean.surname=='zaitseva'].weight.values[0] = patients_clean.weight.min()*2.20462

___Test___

In [127]:
patients_clean.loc[patients_clean.surname=='zaitseva', 'weight'] = patients_clean.weight.min()*2.20462

In [132]:
# 48.8 shouldn't be the lowest anymore
patients_clean.weight.sort_values().head(10)

459    102.100000
335    102.700000
74     103.200000
317    106.000000
171    106.500000
51     107.100000
210    107.585456
270    108.100000
198    108.500000
48     109.100000
Name: weight, dtype: float64